# Bringing the WDL Workflow to AnVIL

Our first step is locating the workflow we want to use. Our goal is to create a subset version of existing genomic data files.

## Find the Workflow on Dockstore

Navigate to the WORKFLOWS tab of your workspace. Select "Find a Workflow". 

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_0")
```

When the dialog box opens, select "Dockstore" on the bottom left. This will take you to the [Dockstore search page](https://dockstore.org/search?descriptorType=WDL&entryType=workflows&searchMode=files).

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_5")
```

Enter `fastq_subsample` in the search box. Select the workflow "fhdsl/AnVIL_WDLs/fastq_subsample".

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_21")
```

## Send the Workflow to AnVIL

On the right side "Launch with" menu, select "AnVIL". This will open a new tab.

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_26")
```

Select the appropriate workspace where you want the workflow to appear. Select "IMPORT".

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_31")
```

You should now be back on AnVIL!

# Set Up the Workflow

## Set Up Inputs

Use the "SELECT DATA" button to select the samples (rows) you want to subset. You can select all or some samples.

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_45")
```

Indicate which columns in the DATA tab are used as workflow inputs.

The first workflow input should be a fastq or zipped fastq file. The workflow calls this input `fastqgz_file_read_1`. Under "Attribute" select the column that contains a link to the first set of reads. 

:::{.notice}
For single-end sequencing, the `fastqgz_file_read_1` input is the only file containing sequencing reads in your data.

For paired-end sequencing, the `fastqgz_file_read_1` input is the first of two read files.
:::

In this example, the column with the fastq file link is called "read1". It will look like "this.read1" under "Attribute".

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_50")
```

Select additional inputs. 

- **Required**: In this example, we've selected "sample_id" as the column containing the name of the sample. This names the output file appropriately. 

- *Optional*: "read2" indicates the second set of reads in our paired-end sequencing approach. Skip this if you have single-end reads.

- *Optional*: Indicate how many reads you want in your subsample file. In this example, we wanted 20,000 reads. (Default: 10,000)

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_55")
```

## Set Up Outputs

Workflow outputs are written to a Google Bucket. Setting up the workflow outputs creates links to these outputs inside the DATA in our workspace, making them easier to locate. 

Select the "OUTPUTS" tab. Select "Use defaults" to use the default output column naming schema. 

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_60")
```

Click "SAVE".

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_66")
```

# Run the Workflow

## Start the Run

Once you have saved the inputs and outputs, click on RUN ANALYSIS.

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_76")
```

## Monitor the Run

Navigate to the JOB HISTORY tab. You can see your most recent submissions in table form. Click on the most recent submission. Notice how each sample gets its own job. This keeps the whole process speedy!

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_81")
```

## Inspect the Run Results

You should see the status change to "Succeeded" if everything completed correctly.

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_86")
```

:::{.notice}
After 24 hours, you can also see the costs associated with each run under "Run Cost".
:::

## Inspecting Other Run Files

It can be helpful to look at intermediate files on Google Cloud Platform, especially if runs did not complete successfully. You can view these files by clicking on the folder icon for "Execution directory".

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_92")
```

For each run, you can see a number of associated files, including the output `.fq` files and log files.

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_97")
```

Click on `stdout` and/or `stderr` and "DOWNLOAD" to view the terminal output.

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_109")
```

Here is what your output in `stdout` might look like.

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_125")
```

## Confirming Results in the DATA tab

We can see that the subsample files have been linked in the DATA table "sample".

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_139")
```

In our example, we have a mixture of single-end and paired-end reads, so only the paired-end samples get a second file.

```{r, fig.align='center', echo = FALSE, fig.alt= "", out.width = '100%'}
ottrpal::include_slide("https://docs.google.com/presentation/d/1zUiupDXMZlZEIditsN5igggOim9qIrR9Ckf4N8fSAB8/edit#slide=id.g2794677b0ab_0_144")
```

You should now be able to use the subset files for downstream applications!